src/client/Microsoft.Identity.Client/ManagedIdentity/ImdsV2ManagedIdentitySource.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Http.Retry;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal class ImdsV2ManagedIdentitySource : AbstractManagedIdentity
+    {
+        private const string CsrMetadataPath = "/metadata/identity/getPlatformMetadata";
+
+        private static CsrMetadata _csrMetadata;
+
+        public static async Task<bool> GetCsrMetadataAsync(RequestContext requestContext)
+        {
+            string uami = null;
+            switch (requestContext.ServiceBundle.Config.ManagedIdentityId.IdType)
+            {
+                case AppConfig.ManagedIdentityIdType.ClientId:
+                    uami = requestContext.ServiceBundle.Config.ManagedIdentityId.UserAssignedId;
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 supports user-assigned client id.");
+                    break;
+
+                case AppConfig.ManagedIdentityIdType.ResourceId:
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 doesn't support user-assigned resource id. Either provide a client id, or use a system-assigned managed identity.");
```

Comment:
Very suprised to see this. I thought MSIv2 is a drop-in replacement for MSIv1. @gladjohn ?

Reply:
This is directly from Raghavendra.

Reply:
We need to support `ResourceId`, else it is a breaking change. As of now per MSI V2 design there is no breaking changes. As Bogdan mentioned this is a drop-in replacement for MSIv1

Reply:
So we should support `ObjectId` as well?

Reply:
done